### ISSUE TITLE
Link to keyscoped topic not properly resolved.

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2538]
I'm attaching a ZIP sample.
[improperLinkToKeyScopedTopic.zip](https://github.com/dita-ot/dita-ot/files/631027/improperLinkToKeyScopedTopic.zip)

So, the DITA Map looks like this:
```xml
<topicgroup keyscope="product1_scope">
  ...........
 <topicref keys="archiving" href="product1/c_archiving_mod.dita"/>
</topicgroup>
<topicgroup keyscope="product2_scope">
 ...............
 <topicref keys="archiving" href="product1/c_archiving_mod.dita"/>
</topicgroup>
<topicref href="linking.dita"/>
```

Because "c_archiving_mod.dita" is referenced in two key scopes, we will have in the output file two HTMLs for it: "c_archiving_mod.html" and "c_archiving_mod-1.html". And that's good.
The "linking.dita" should generate links to both those HTML files using fully qualified key scoped:
```xml
<p>Links <xref keyref="product1_scope.archiving"
/> to <xref keyref="product2_scope.archiving"
 />.</p>
```
But in the generated "linking.html" you will have both links point to the same HTML file (the one generated from the first context).

I looked a little bit in the code, the method `org.dita.dost.writer.KeyrefPaser.processElement(Attributes)` resolves key references based on the current "KeyScope" map. Unfortunately the map contains these mappings:
```
product1_scope.archiving=product1_scope.archiving=product1/c_archiving_mod.dita
product2_scope.archiving=product2_scope.archiving=product1/c_archiving_mod.dita
```
although in my opinion the "product2_scope.archiving" should have been re-written (possibly in the org.dita.dost.module.KeyrefModule.execute) to point to "product1/c_archiving_mod-1.dita" because the DITA Map has already been re-written to do so.

Might also be related with #2523

**Comment by raducoravu** [comment]
Somehow when the "org.dita.dost.module.KeyrefModule.walkMap" updates the @href's in the DITA Map the key scopes would also need to be updated to point to the new hrefs... Not sure how this could be done though.

**Comment by jelovirt** [comment]
I'm currently debugging this. It's kinda tricky to find the best stage when to rewrite the key targets. I think the only correct location is after all `ResolveTask` jobs have been generated. Then we know all the rewrites that are needed in for all scopes.

**Comment by raducoravu** [comment]
Or have a single iteration over the map DOM element which would produce the key scopes + would modify the map nodes with the unique href attribute values so the key scopes would be produced from the beginning with references from the keys to the unique href values.

**Comment by jelovirt** [comment]
Since a single keyscope may contain multiple keys that need to be rewritten, we need to have the whole scope defined before we process the duplicates. Thus we can't read key scope at the same time as we define renames.

**Comment by jelovirt** [comment]
@raducoravu #2539 has a partial fix to the issue. Scope prefixed keyrefs don't still work with duplicates, but at least plain keyrefs would work. Would you be ok if that was included in 2.4.1 and a new issue for the scope prefixed references would be created and fixed later?

**Comment by raducoravu** [comment]
I'm not sure how I could test your fix, I'm not an expert GIT user, I see you have not yet merged it back to the 2.4.1 branch.
So the problem described in the sample I posted would still not be solved, right (because it uses scoped keyrefs)? Then maybe you should add an issue with samples for the issue that you solved...

**Comment by jelovirt** [comment]
The scope prefixed links in `linking.dita` are not processed correctly; links in `product1/c_archiving_mod.dita` are output correctly.

**Comment by xephon2** [comment]
@raducoravu , you have to add Jarno's dita-ot fork as a remote repository:

```bash
git remote add jelovirt https://github.com/jelovirt/dita-ot.git
```

Then you have to checkout his feature branch `feature/keyscope_target_rewrite`.

```bash
git checkout -b keyscope_target_rewrite jelovirt/feature/keyscope_target_rewrite
```

I've not tested this, but it should work.

**Comment by raducoravu** [comment]
Thanks @xephon2, I succeeded now.
@jelovirt The 2 keyrefs in the "c_archiving_mod.dita" were properly resolved from the very beginning in both contexts. I tested your fix by creating another topic which has a link using 'keyref="archiving"' and using it in both key scopes and indeed your fix now properly resolves the targets. 
As for the initial samples, I don't think your fix influences them in any way. So I understand what you did, it is better in this regard but I was really looking after having keyscoped keyreffed links working in this particular issue. So if you want I can add another issue with a sample exhibiting exactly the problem that you fixed.

As a possible implementation approach (I'm not fully understanding the code so I am probably wrong but this never stopped me) maybe the "KeyScope" objects could have stored from the very beginning, when they were constructed by iterating the DOM map a weak hash map having as keys the topicref DOM nodes and as values the keys that it contributed to the "keyDefinition" map. In this way, when you iterate again through the DOM map and change hrefs, you know exactly what mappings to change in the keyscope objects. And make modifications directly to the existing keyscope objects, stop considering them as immutable, at least in this step.

**Comment by lief-erickson** [comment]
+1 to @raducoravu's scenario. I do this type of scoped key linking (also sometimes in conjunction with branch filtering) in my files, both as inline xrefs as well as reltable links.

**Comment by raducoravu** [comment]
@jelovirt I added as a proposed solution a pull request suggesting a change that I will probably implement in the patches we have for the DITA OT:
https://github.com/dita-ot/dita-ot/pull/3344
but I have not yet tested the change on lots of various situations.

**Comment by raducoravu** [comment]
Fixed in DITA OT 3.4 with pull request #3344
